sc/src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java

```diff
@@ -102,15 +102,15 @@ public GeneSearchDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate,
             "WHERE experiment_accession=:experiment_accession " +
                 "AND gene_id=:gene_id " +
                 "AND marker_probability<:threshold " +
-                "AND ((k is NULL OR k=:preferred_K) " +
+                "AND (k=:preferred_K " +
```

Comment:
Hmm I think that was meant to be `:preferred_k IS NULL` (see [here](https://stackoverflow.com/questions/21062148/how-to-query-postgres-on-optional-params))

Reply:
I’m not entirely sure about that... We want this query to return the following:
1. If we pass a valid `preferredK`, two rows: one with a matching *k* and the cluster ID and another with the *k* and cluster ID with the lowest *p*-value (they could be the same).
2. If we pass `null` (or 0): only the *k* and cluster ID with the lowest *p*-value.

If we have `:preferred_k IS NULL`, in the second case we would get all the rows where the *p*-value is below the threshold since the third clause would evaluate tot true.

Unless I’m missing something here...

Reply:
Yeah, I think that's correct... Yeah maybe the two statements used to be AND-ed together? I don't know... Oh, something else to note is that this statement won't necessarily return just 2 results... There are cases where multiple k's have the same minimum probability, it will return all of them... Otherwise how do you choose which `k` to discard?

Reply:
I guess that the most probable case of that happening is if the *p*-value is 0 (otherwise we’re being astronomically unlucky, but yes, it’s possible). Which k should we pick in that case? The more I think about this, the more I lean towards making two SQL queries...

Reply:
It's not just when the p-value is 0 (though it's close). Gene `ENSG00000167748` in `E-MTAB-5061` has the same (minimum) p-value for k = 7, 8, 9, 10, 11, 12, 13, 17... Unless the query is wrong.

Reply:
I see. It turns out that this happens when it’s the same cluster, which makes sense:
• https://wwwdev.ebi.ac.uk/gxa/sc/experiments/E-GEOD-81547/Results?geneId=ENSG00000167748&k=13&clusterId=[4]
• https://wwwdev.ebi.ac.uk/gxa/sc/experiments/E-GEOD-81547/Results?geneId=ENSG00000167748&k=14&clusterId=[8]

We’ll need to discuss which k is then preferred.